arangod/Replication2/Storage/WAL/FileWriterImpl.cpp

```diff
@@ -0,0 +1,79 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2023 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Manuel Pöter
+////////////////////////////////////////////////////////////////////////////////
+
+#include "FileWriterImpl.h"
+
+#include <fcntl.h>
+#include <cstring>
+
+#include "Assertions/ProdAssert.h"
+#include "Basics/Exceptions.h"
+#include "Basics/voc-errors.h"
+#include "Logger/LogMacros.h"
+#include "Replication2/Storage/WAL/FileReaderImpl.h"
+
+namespace arangodb::replication2::storage::wal {
+
+FileWriterImpl::FileWriterImpl(std::string path) : _path(std::move(path)) {
+  _file = ::open(_path.c_str(), O_CREAT | O_RDWR | O_APPEND, S_IRUSR | S_IWUSR);
```

Comment:
I have added `O_CLOEXEC`. I have added code to fsync the directory entries when we create them so we probably don't need to do anything else here, agreed?

Reply:
No, that doesn't suffice. It's not about syncing the directory entry after it has been created.

After a new hard link is created in a directory, the directory needs to be synced again: Otherwise, the file itself might not be part of the directory after a crash, and then it's no use that its data was synced to disk.